### PR TITLE
Add LayerNorm1P, weight decay for LN and unscaled initialization

### PR DESCRIPTION
**Pull request by mikolajblaz**
# What does this PR do ?

Port Megatron features to match GPT3 convergence

**Collection**: NLP/Megatron

# Changelog 
- Add LayerNorm1P class implementing numerically stable LayerNorm with formula: `(self.weight + 1) * x + self.bias` instead of `self.weight * x + self.bias`
- add option to use weight decay also on LayerNorm weights
- add option to use unscaled residuals initialization

All 3 options are exposed with new hydra flags. With default flag values, the **model behaviour is unchanged**.

# Usage
Override these flags to enable respective features described above:
- model.normalization=layernorm1p
- model.do_layer_norm_weight_decay=1
- model.use_scaled_init_method=0

Each of the flags work independently from each other

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

### CODE REVIEW

**Comment by mikolajblaz** [comment]
Some of the TTS and ASR related tests fail, but they fail also on r1.11.0 branch, so it's unrelated to this PR

**Review by ericharper — DISMISSED** [review]
LGTM. Thanks!

**Comment by ericharper** [comment]
Why is this going to r1.11.0 ? That branch is code frozen and should only be accepting bug fixes and documentation updates.

File: nemo/collections/nlp/modules/common/megatron/layer_norm_1p.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    from apex.transformer.layers.layer_norm import FastLayerNorm
+
+    HAVE_APEX = True
+except (ImportError, ModuleNotFoundError):
+    HAVE_APEX = False
+
+
+if HAVE_APEX:
+    # TODO: use Apex implementation
+    class LayerNorm1P(FastLayerNorm):
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/modules/common/megatron/layer_norm_1p.py:29]: Can you please add a docstring here indicating that this is a stability trick for layernorm with bf16?

**Review by ericharper — APPROVED** [review]
Re-approving.

File: nemo/collections/nlp/modules/common/megatron/layer_norm_1p.py (hunk 1 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/megatron/layer_norm_1p.py:29]: Sorry I just saw this after merging.

**Pull request merged**